### DISCUSSION TITLE
Atari 5200 Diagnostic Cart Linker cfg file Thoughts

### DISCUSSION THREAD

**Discussion by baktragh** [cc65/cc65#1371] [Ideas]
Hello,

I was thinking of contributing a linker .cfg file to create atari5200 diagnostic carts (atari5200-diagcart.cfg). However, I am now wondering of a real value of such contribution. The .cfg file itself is an easy task.

What concerns me are the library functions. A diagnostic cart code is entered right after power-up with no VBI and DLI.
Therefore, some of the library functions will not work out of the box (e.g. all conio functions, as they expect the ANTIC and GTIA to be ready).

A diagnostic cart is typically created for two purposes:
1. Bypass the Atari rainbow logo and its Y2K problem. Useful for home brew games. More likely.
2. Create a real cart for system diagnostics. Less likely.

So I see three possibilities:

1. Do not ship any .cfg file and let developers creating diagnostic carts on their own.
2. Ship the .cfg file and add note in the documentation saying that conio functions will not work until the system is set up for that manually
3. Ship the .cfg file and a modified library (perhaps just .o file) that would make sure conio works always. This solution would require a code that would set up the VBI including its handler, basically duplicating code in OS ROM. Not a trivial task.

Any thoughts on this?

**Comment by oliverschmidt** [comment]
Two thoughts:

* Option 1.) could be improved by writing a Wiki page explaining what needs to be known.

* Option 3.) could be likely/hopefully realized with an "extra" .o file. Look out for `extra` subdirectories in the `libsrc` tree for exsisting use cases.

**Comment by groessler** [comment]
I don't see the need for a different cfg file.

Am I correct that a diag card is identified by 0xFF at 0xBFFD? Then you can create one with the current cfg by providing your own "CARTYEAR" segment:

```
.export         __CART_YEAR__: absolute = 1
.segment        "CARTYEAR"
                .byte   255,255
```

regards,
chris

**Reply by baktragh** [reply]
That is indeed correct and will work.
Is it possible to get rid of the CARTNAME segment (which is not needed for diagnostic carts) without modification of the .cfg file?
Just curious, not that those few wasted CARTNAME bytes would matter too much anyway.

**Reply by groessler** [reply]
Honestly, I think if somebody manages to get that tight on memory that these 20 bytes are important, he or she can also be bothered to edit the cfg file.

**Reply by baktragh** [reply]
I agree

**Comment by polluks** [comment]
Indeed, we don't want inflation of configuration files. However a new section in atari.sgml about Diagnostic Carts may be useful.

**Comment by baktragh** [comment]
So far it seems the public opinion is against a new .cfg file. 
Perhaps a smal doc update would be sufficient - it can include the sample CARTYEAR segment mentioned above.

**Reply by groessler** [reply]
Sure.
If you are doing it, could you also mention that the program name in the splash screen can also be changed?
Supply your own version of this:

```
.export         __CART_NAME__: absolute = 1
.macpack        atari
.segment        "CARTNAME"
                scrcode "   cc"
                .byte   '6' + 32, '5' + 32      ; use playfield 1
                scrcode " compiled"
```

Default name is just "cc65 compiled".

**Reply by baktragh** [reply]
Yes, that is worth mentioning too.

**Comment by groessler** [comment]
Option 3) of your original proposal would allow to have a splash screen with y2k fix and own copyright message.
I was planning to implement such but haven't come around to do it.

**Reply by baktragh** [reply]
That would mean the .o file would include the following:

1. Ensuring the cartridge is a diagnostic cartridge
2. Setting-up interrupts and interrupt handlers
3. Providing a splash screen that is the same as in the OS ROM (rainbow logo), but with custom copyright text

With points 2 and 3, a decision would have to be made if the code will be fully in the generated cartridge (the interrupt handlers for example), or some portions of the OS ROM will be used (i.e. vector to interrupt handlers will point to OS ROM).

**Comment by groessler** [comment]
I would intend to re-use BIOS code as far as possible to keep the memory overhead in the cartridge low.

And, yes, I'd also like to have a knob to disable the splash screen. It's nice, but getting old and annoying with time (especially the wait time).

**Reply by baktragh** [reply]
I would tend to agree with using OS ROM code. Fortunately, there are not so many alternate OS ROMs.
I know of Altirra OS for 5200, but that one tries to keep maximum compatibility (with phaeron's complaints on 3rd party developers in its source code)

**Reply by groessler** [reply]
I thought there were only two versions, 2-port and 4-port. Didn't know about Altirra version. But maybe it can be made to work with those three.

**Reply by baktragh** [reply]
When I am thinking about this, would it be acceptable to provide such functionality as a library function - i.e something that would be called from main()? I did something like that in my latest game for this console.

```
#include <atari5200.h>
int main() {
  /*Call diagnostic cart standard initialization, splash screen delay 3 seconds*/
  diagcartinit("DOOM 8"," (C) 2026 ID SOFTWARE",3); 
}
```

**Reply by groessler** [reply]
I would call the function something like `splashscreen()` or so, but yes, I think that's a good idea. By default there would be no initial screen, but it can be displayed under program control.

**Comment by groessler** [comment]
Selectable wait time... Pah... :-)

**Comment by groessler** [comment]
I've written something to change name (in splash screen) and year/diagnostic cart in PR #1401. Feel free to extend...
